roxie/ccd/ccdqueue.cpp

```diff
@@ -608,51 +607,112 @@ void SlaveContextLogger::flush()
+        while (subChannel != mySubChannel)
+        {
+            delay += currentDelay[subChannel];
+            subChannel++;
+            if (subChannel == numSubChannels)
+                subChannel = 0;
+        }
+        return delay;
+    }
 
-unsigned getIbytiDelay(unsigned channel, unsigned primarySubChannel)
-{
-    // MORE - adjust delay according to whether it's a retry, whether it was a broadcast etc
-    unsigned mySubChannel = subChannels[channel] - 1;
-    unsigned subChannel = primarySubChannel;
-    unsigned delay = 0;
-    while (subChannel != mySubChannel)
-    {
-        if (isSubChannelAlive(channel, subChannel))
-            delay++;
-        subChannel++;
-        if (subChannel >= numSlaves[channel])
-            subChannel = 0;
-    }
-    return delay * initIbytiDelay;
-}
+    void noteChannelsSick(unsigned subChannel)
+    {
+        while (subChannel != mySubChannel)
+        {
+            unsigned newDelay = currentDelay[subChannel] / 2;
+            if (newDelay < minIbytiDelay)
+                newDelay = minIbytiDelay;
+            currentDelay[subChannel] = newDelay;
+            subChannel++;
+            if (subChannel == numSubChannels)
+                subChannel = 0;
+        }
+    }
 
-void notePrimarySubChannelsDead(unsigned channel, unsigned primarySubChannel)
-{
-    unsigned mySubChannel = subChannels[channel] - 1;
-    unsigned subChannel = primarySubChannel;
-    while (subChannel != mySubChannel)
+    void noteChannelHealthy(unsigned subChannel)
     {
-        liveSubChannels[channel] &= ~(SUBCHANNEL_MASK << (SUBCHANNEL_BITS * subChannel));
-        subChannel++;
-        if (subChannel >= numSlaves[channel])
-            subChannel = 0;
+        currentDelay[subChannel] = initIbytiDelay;
     }
-}
+
+    void init(unsigned channel)
+    {
+        mySubChannel = subChannels[channel]-1;
+        numSubChannels = numSlaves[channel];
+        for (unsigned i = 0; i < numSubChannels; i++)
+            currentDelay[i] = initIbytiDelay;
+    }
+private:
+    unsigned mySubChannel = 0;  // zero-based
+    unsigned numSubChannels = 0;
+    unsigned currentDelay[MAX_SUBCHANNEL] = {0};
+};
+
+static SubChannelInfo subChannelInfo[MAX_CLUSTER_SIZE];
 
 bool otherSlaveHasPriority(const RoxiePacketHeader &h)
 {
     unsigned hash = h.priorityHash();
-    unsigned primarySubChannel = (hash % numSlaves[h.channel]) + 1;
-    unsigned mySubChannel = subChannels[h.channel];
+    unsigned primarySubChannel = (hash % numSlaves[h.channel]);
+    unsigned mySubChannel = subChannels[h.channel] - 1;
+    unsigned theirSubChannel = h.getRespondingSubChannel();
+    // could be coded smarter! Basically mysub - prim < theirsub - prim using modulo arithmetic, I think
     while (primarySubChannel != mySubChannel)
     {
-        unsigned channelMask = SUBCHANNEL_MASK << (SUBCHANNEL_BITS * primarySubChannel);
-        if ((h.retries & ROXIE_RETRIES_MASK) == channelMask)
+        if (primarySubChannel == theirSubChannel)
```

Comment:
still trying to understand this change...

Reply:
It was supposed to make this code clearer! Net effect will be the same.